README
	
= Introduction =

Please visit the following websites where you can find more
information:

Project Honey Pot website
http://www.projecthoneypot.org/?rf=39586

Extension wiki-httpbl
http://code.google.com/p/wiki-httpbl/

= Contact =

Author: Mario Oyorzabal Salgado
Blog: http://blog.tuxsoul.com
E-mail: <tuxsoul@tuxsoul.com>

= Bugs =

If you find any bugs or have any idea for a feature can be usefull to this 
extension, you can leave it in the bugtrack:

http://code.google.com/p/wiki-httpbl/issues/list

= License =

GNU General Public License v2
http://www.gnu.org/licenses/old-licenses/gpl-2.0.html

= Requirements =

  * Key from "Project Honey Pot"
  * Quicklink from "Project Honey Pot"
  * Mediawiki version 1.6 +
  * PHP4 +

= Features =

  * Use Http:BL to detect spammer's
  * Ability to redirect into a trap if the visitor is a spammer
  * Ability to insert addresses traps in the html code of
    Mediawiki

= Details =

  * First download and decompresses the file extension of this
  * Rename the directory the this extension to "wiki-httpbl"
  * Upload the "wiki-httpbl" to the directory of extensions you wiki, which is 
    called "extensions"
  * Open the file named "LocalSettings.php" which is located in the
    Root directory of your site, after adding the following two lines to
    End of the file:

	##### Include Http:bl
	# Key of access to Http:BL (Project Honey Pot)
	$wgWikiHttpBLKey = '------------';
	# QuickLink (Project Honey Pot)
	$wgWikiHttpBLTrap = '';
	# Word to work whith Quicklink (url trap)
	$wgWikiHttpBLTrapWord = 'mediawiki';
	include( './extensions/wiki-httpbl/httpbl.php' );

  * Save the file and enters your website to see if it is not showing any
    Error
 
Lastly and can only wait to block the spammer's coming to you Website.

= Donation =

Please consider helping through donations, for more information visit the 
following link:

Donate
http://tinyurl.com/2zhdhg

Extension wiki-httpbl
http://code.google.com/p/wiki-httpbl/


Cheers :-).


CHANGELOG
================================================================================
CHANGELOG
================================================================================

--------------------------------------------------------------------------------
2008-02-01  -  v0.2.2
--------------------------------------------------------------------------------
Whith this release you can configure Http:BL from "LocalSettings.php", for more
easy install and future updates.

* Delete: "config.httpbl.php"
* Rename: "core.httpbl.php" to "httpbl.core.php"
* Rename: "object.httpbl.php" to "httpbl.object.php"
* Add: "$wgWikiHttpBLKey", set key access to Http:BL (Project Honey Pot)
* Add: "$wgWikiHttpBLTrap", set quicklink (url trap)
* Add: "$wgWikiHttpBLTrapWord", set word to work whith quicklink
* Update: README, README-ES

--------------------------------------------------------------------------------
2008-01-31  -  v0.1.1
--------------------------------------------------------------------------------
* Fix bug #1: "wiki-httpbl stop others extensions using the same wgHook ...."
* Change: $wgHook['ParserBeforeTidy'] to $wgHook['ParserAfterTidy']
* Change: version to v0.1.1

--------------------------------------------------------------------------------
2008-01-20  -  v0.1
--------------------------------------------------------------------------------
* Core version '0.1'
* First version of this extension
* Detect spammer's using http:bl from 'Project Honey Pot'
* Can redirect spammer to one address trap
* Use one word with address trap
* Insert three or more address trap in the html code

================================================================================


core.httpbl.php
<?php
/*
Implementation of API from "Project Honey Pot"
http://www.projecthoneypot.org/httpbl_api

Author: Mario Oyorzabal Salgado
Blog:   http://blog.tuxsoul.com/
E-mail: <tuxsoul@tuxsoul.com>
Web:    http://code.google.com/p/wiki-httpbl/
Donate: http://tinyurl.com/2zhdhg

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

// can't execute directly this file
// you are lost ?
if (!defined('HTTPBL_CWD')) {
	die ('You are lost ?');
}

// main class object
class coreHttpBL 
{
	var $coreHttpBLIp;
	var $coreHttpBLError;

	var $coreHttpBLConfVersion;
	var $coreHttpBLConfDnsbl;
	var $coreHttpBLConfDays;

	function coreHttpBL () {
		// set flag error to cero
		$this->coreHttpBLError = 0;
	
		// config
		// version of this core
		$this->coreHttpBLConfVersion = '0.1';

		// spammer days since last activity
		$this->coreHttpBLConfDays = '15';

		// define query to dnsbl
		$this->coreHttpBLConfDnsbl = '{key}.{ip}.dnsbl.httpbl.org';
	}

	// return version of this core
	function coreHttpBLVersion () {
		return $this->coreHttpBLConfVersion;
	}

	// start the search to spammer's
	function coreHttpBLStart () {
		// detect ip
		$this->coreHttpBLDetectIp();

		// reverse-octet ip address
		// change data of query
		$host = $this->coreHttpBLDnsblPrepareQuery();

		// make query and get response
		$response = $this->coreHttpBLDnsblQuery($host);

		// Is spammer ?
		if($this->coreHttpBLIsSpammer($response) == 1) {
			// Get the trap
			$this->coreHttpBLShowTrap();
		}
	}

	// try to detect the remote ip
	function coreHttpBLDetectIP () {
		$ip = $_SERVER['REMOTE_ADDR'];

		// only for test
		// $ip = '127.1.1.1';

		$this->coreHttpBLIp = $ip;
	}

	// change data of query to dnsbl
	function coreHttpBLDnsblPrepareQuery () {
		// reverse-octet ip format
		$reverseIp = implode('.', array_reverse(explode('.', $this->coreHttpBLIp)));

		// prepare query to dnsbl
		$string = $this->coreHttpBLConfDnsbl;
		$string = ereg_replace('{key}', HTTPBL_KEY, $string);
		$string = ereg_replace('{ip}', $reverseIp, $string);

		return $string;
	}

	// make the query to dnsbl server
	function coreHttpBLDnsblQuery ($host) {
		$response = gethostbyname($host);

		if($response == $host)
			$this->coreHttpBLError = 1;

		return $response;
	}

	// is spammer ?
	function coreHttpBLIsSpammer ($response) {
		if($this->coreHttpBLError == 0) {
			$data = explode('.', $response);
			$spammer = 0;

			// the first octect always is 127
			if((int)$data[0] == 127) {
				// spammer days since last activity
				if((int)$data[1] <= $this->coreHttpBLConfDays) {
					// only check the type of spammer
					if((int)$data[3] > 0) {
						$spammer = 1;
					}
				}		
			}
			else
				$this->coreHttpBLError = 1;

			return $spammer;
		}

		return 0;
	}

	// redirect to the trap
	function coreHttpBLShowTrap () {
		if($this->coreHttpBLError == 0) {
			// FIXME: HTTPBL_TRAP is set correctly ?
			if(strlen(HTTPBL_TRAP) > 0) {
				header('Location: ' . HTTPBL_TRAP);
				die();
			}
		}
	}

	// return url to trap in html
	// FIXME: Need more work here :)
	function coreHttpBLGetUrlTrap () {
		$text = '<!-- begin links ' . HTTPBL_TRAP_WORD . ' -->';

		// FIXME: HTTPBL_TRAP is set correctly ?
		if(strlen(HTTPBL_TRAP) > 0) {
			$url = array (
					'<a href="{url}"><!-- {word} --></a>',
					'<a href="{url}" style="display: none;">{word}</a>',
					'<div style="display: none;"><a href="{url}">{word}</a></div>',
					'<a href="{url}"></a>',
					'<!-- <a href="{url}">{word}</a> -->',
					'<div style="position: absolute; top: -250px; left: -250px;"><a href="{url}">{word}</a></div>',
					'<a href="{url}"><span style="display: none;">{word}</span></a>',
				);

			for($x = 0; $x <= 2; $x++) {
				$temp = $url[rand(0, 6)];
				$temp = ereg_replace('{url}', HTTPBL_TRAP, $temp);
				$temp = ereg_replace('{word}', HTTPBL_TRAP_WORD, $temp);

				$text .= $temp;
			}

			$text .= '<!-- end links '. HTTPBL_TRAP_WORD . ' -->';
		}

		return $text;
	}
}
?>


httpbl.core.php
<?php
/*
Implementation of API from "Project Honey Pot"
http://www.projecthoneypot.org/httpbl_api

Author: Mario Oyorzabal Salgado
Blog:   http://blog.tuxsoul.com/
E-mail: <tuxsoul@tuxsoul.com>
Web:    http://code.google.com/p/wiki-httpbl/
Donate: http://tinyurl.com/2zhdhg

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

// can't execute directly this file
// you are lost ?
if (!defined('HTTPBL_CWD')) {
	die ('You are lost ?');
}

// main class object
class coreHttpBL 
{
	// config values
	var $key, $url, $word;

	var $coreHttpBLIp;
	var $coreHttpBLError;

	var $coreHttpBLConfVersion;
	var $coreHttpBLConfDnsbl;
	var $coreHttpBLConfDays;

	function coreHttpBL () {
		// set flag error to cero
		$this->coreHttpBLError = 0;
	
		// config
		// version of this core
		$this->coreHttpBLConfVersion = '0.2';

		// spammer days since last activity
		$this->coreHttpBLConfDays = '15';

		// define query to dnsbl
		$this->coreHttpBLConfDnsbl = '{key}.{ip}.dnsbl.httpbl.org';
	}

	// return version of this core
	function coreHttpBLVersion () {
		return $this->coreHttpBLConfVersion;
	}

	// start the search to spammer's
	function coreHttpBLStart () {
		// detect ip
		$this->coreHttpBLDetectIp();

		// reverse-octet ip address
		// change data of query
		$host = $this->coreHttpBLDnsblPrepareQuery();

		// make query and get response
		$response = $this->coreHttpBLDnsblQuery($host);

		// Is spammer ?
		if($this->coreHttpBLIsSpammer($response) == 1) {
			// Get the trap
			$this->coreHttpBLShowTrap();
		}
	}

	// try to detect the remote ip
	function coreHttpBLDetectIP () {
		$ip = $_SERVER['REMOTE_ADDR'];

		// only for test
		// $ip = '127.1.1.1';

		$this->coreHttpBLIp = $ip;
	}

	// change data of query to dnsbl
	function coreHttpBLDnsblPrepareQuery () {
		// reverse-octet ip format
		$reverseIp = implode('.', array_reverse(explode('.', $this->coreHttpBLIp)));

		// prepare query to dnsbl
		$string = $this->coreHttpBLConfDnsbl;
		$string = ereg_replace('{key}', $this->key, $string);
		$string = ereg_replace('{ip}', $reverseIp, $string);

		return $string;
	}

	// make the query to dnsbl server
	function coreHttpBLDnsblQuery ($host) {
		$response = gethostbyname($host);

		if($response == $host)
			$this->coreHttpBLError = 1;

		return $response;
	}

	// is spammer ?
	function coreHttpBLIsSpammer ($response) {
		if($this->coreHttpBLError == 0) {
			$data = explode('.', $response);
			$spammer = 0;

			// the first octect always is 127
			if((int)$data[0] == 127) {
				// spammer days since last activity
				if((int)$data[1] <= $this->coreHttpBLConfDays) {
					// only check the type of spammer
					if((int)$data[3] > 0) {
						$spammer = 1;
					}
				}		
			}
			else
				$this->coreHttpBLError = 1;

			return $spammer;
		}

		return 0;
	}

	// redirect to the trap
	function coreHttpBLShowTrap () {
		if($this->coreHttpBLError == 0) {
			// FIXME: $this->url is set correctly ?
			if(strlen($this->url) > 0) {
				header('Location: ' . $this->url);
				die();
			}
		}
	}

	// return url to trap in html
	// FIXME: Need more work here :)
	function coreHttpBLGetUrlTrap () {
		$text = '<!-- begin links ' . $this->word . ' -->';

		// FIXME: $this->url is set correctly ?
		if(strlen($this->url) > 0) {
			$url = array (
					'<a href="{url}"><!-- {word} --></a>',
					'<a href="{url}" style="display: none;">{word}</a>',
					'<div style="display: none;"><a href="{url}">{word}</a></div>',
					'<a href="{url}"></a>',
					'<!-- <a href="{url}">{word}</a> -->',
					'<div style="position: absolute; top: -250px; left: -250px;"><a href="{url}">{word}</a></div>',
					'<a href="{url}"><span style="display: none;">{word}</span></a>',
				);

			for($x = 0; $x <= 2; $x++) {
				$temp = $url[rand(0, 6)];
				$temp = ereg_replace('{url}', $this->url, $temp);
				$temp = ereg_replace('{word}', $this->word, $temp);

				$text .= $temp;
			}

			$text .= '<!-- end links '. $this->word . ' -->';
		}

		return $text;
	}
}

?>


config.httpbl.php
<?php
/*
Implementation of API from "Project Honey Pot"
http://www.projecthoneypot.org/httpbl_api

Author: Mario Oyorzabal Salgado
Blog:   http://blog.tuxsoul.com/
E-mail:	<tuxsoul@tuxsoul.com>
Web: 	http://code.google.com/p/wiki-httpbl/
Donate: http://tinyurl.com/2zhdhg

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

// you can't execute directly this file
// you are lost ?
if (!defined('HTTPBL_CWD')) {
	die ('You are lost ?');
}

////////// Configuration
////////// You can edit all values ;-)

///// Key to access dnsbl service from "Project Honey Pot"
///// All Access Keys are 12-characters in length, lower case, and contain only alpha characters (no numbers)
///// You need a key ?, register here: http://www.projecthoneypot.org?rf=39586
define('HTTPBL_KEY', '------------');

////////// Url to honeypot trap for redirect malicious visitors
///// If you have install a trap or you can get this in "Project Honey Pot" -> "QuickLink"
///// This value is important to catch more spammers ;-)
define('HTTPBL_TRAP', '');

///// Is important change this value to any word, for have usefull url trap's
define('HTTPBL_TRAP_WORD', 'mediawiki');

?>


httpbl.object.php
<?php
/*
Implementation of API from "Project Honey Pot"
http://www.projecthoneypot.org/httpbl_api

Author: Mario Oyorzabal Salgado
Blog:   http://blog.tuxsoul.com/
E-mail: <tuxsoul@tuxsoul.com>
Web:    http://code.google.com/p/wiki-httpbl/
Donate: http://tinyurl.com/2zhdhg

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

// can't execute directly this file
// you are lost ?
if (!defined('HTTPBL_CWD')) {
	die('You are lost ?');
}

// include core file
include (HTTPBL_CWD . '/httpbl.core.php');

// main class object
class objectHttpBL extends coreHttpBL
{
	// start the search to spammer's
	function httpblStart (&$data) {
		// set the key to access 'Project Honey Pot'
		$this->key = $data[0];
		$this->url = $data[1];
		$this->word = $data[2];

		$this->coreHttpBLStart();
	}

	// get url to trap
	function httpblGetUrlTrap (&$url, &$word) {
		$this->url = $url;
		$this->word = $word;

		return $this->coreHttpBLGetUrlTrap();
	}

	// return information from this code
	function httpblInfo ($type) {
		$description  = 'This code implement the API from "Project Honey Pot", to detect Spammers, Harvesters';
		$description .= ' and other malicious visitors. Consider donate to this extension';

		$values = array (
					'name' => 	 'Http:BL',
					'version' => 	 $this->coreHttpBLVersion(),
					'author' =>	 'Mario Oyorzabal Salgado',
					'description' => $description,
					'url' =>	 'http://code.google.com/p/wiki-httpbl/',
				);

		if (isset($values[$type]))
			return $values[$type];
		else
			return 'Null';
	}
}

?>


object.httpbl.php
<?php
/*
Implementation of API from "Project Honey Pot"
http://www.projecthoneypot.org/httpbl_api

Author: Mario Oyorzabal Salgado
Blog:   http://blog.tuxsoul.com/
E-mail: <tuxsoul@tuxsoul.com>
Web:    http://code.google.com/p/wiki-httpbl/
Donate: http://tinyurl.com/2zhdhg

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

// can't execute directly this file
// you are lost ?
if (!defined('HTTPBL_CWD')) {
	die('You are lost ?');
}

// include config file
include (HTTPBL_CWD . '/config.httpbl.php');

// include core file
include (HTTPBL_CWD . '/core.httpbl.php');

// main class object
class objectHttpBL extends coreHttpBL
{
	// load configuration
	function httpbl () {

	}

	// start the search to spammer's
	function httpblStart () {
		$this->coreHttpBLStart();
	}

	// get url to trap
	function httpblGetUrlTrap () {
		return $this->coreHttpBLGetUrlTrap();
	}

	// return information from this code
	function httpblInfo ($type) {
		$description  = 'This code implement the API from "Project Honey Pot", to detect Spammers, Harvesters';
		$description .= ' and other malicious visitors. Consider donate to this extension';

		$values = array (
					'name' => 	 'Http:BL',
					'version' => 	 $this->coreHttpBLVersion(),
					'author' =>	 'Mario Oyorzabal Salgado',
					'description' => $description,
					'url' =>	 'http://code.google.com/p/wiki-httpbl/',
				);

		if (isset($values[$type]))
			return $values[$type];
		else
			return 'Null';
	}
}
?>


httpbl.php
<?php
/*
Implementation of API from "Project Honey Pot"
http://www.projecthoneypot.org/httpbl_api

Author: Mario Oyorzabal Salgado
Blog:   http://blog.tuxsoul.com/
E-mail: <tuxsoul@tuxsoul.com>
Web:    http://code.google.com/p/wiki-httpbl/
Donate: http://tinyurl.com/2zhdhg

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

// Check to run with mediawiki
if (!defined('MEDIAWIKI')) {
	die ('You are lost ?');
}

// Extension version, not core version, is other stuff
$httpblExtensionVersion = "2";

// Define path of extension
define('HTTPBL_CWD', dirname(__FILE__));

// Include class to work dnsbl of httpbl
include(HTTPBL_CWD . '/httpbl.object.php');

// Instance object
$httpbl = new objectHttpBL();

// Execute httpbl filter
function httpbl_mediawiki_start () {
	global $httpbl, $wgWikiHttpBLKey, $wgWikiHttpBLTrap, $wgWikiHttpBLTrapWord;

	$data = array (
			$wgWikiHttpBLKey,
			$wgWikiHttpBLTrap,
			$wgWikiHttpBLTrapWord
		);

	$httpbl->httpblStart($data);
}

// Execute httpbl text
function httpbl_mediawiki_text (&$parser, &$text) {
	global $httpbl, $wgWikiHttpBLTrap, $wgWikiHttpBLTrapWord;

	$text = $httpbl->httpblGetUrlTrap($wgWikiHttpBLTrap, $wgWikiHttpBLTrapWord) . $text;
	return true;
}

// Extension info
$wgExtensionCredits['other'][] = array(
       	'name' => $httpbl->httpblInfo('name'),
        'version' => $httpbl->httpblInfo('version') . '.' . $httpblExtensionVersion,
       	'author' => $httpbl->httpblInfo('author'),
        'description' => $httpbl->httpblInfo('description'),
        'url' => $httpbl->httpblInfo('url'),
);

$wgHooks['ParserAfterTidy'][] = 'httpbl_mediawiki_text';
$wgExtensionFunctions[] = 'httpbl_mediawiki_start';

?>
